In [50]:
#enables chart rendering
%matplotlib inline

# First 10 Minutes 

Welcome to your first 10 minutes with Pyrugga. In this tutorial you will learn how to convert a Superscout XML into a Match object analyse a game of Rugby. 

The first step is to import Pyrugga library. This is as simple us typing 

In [51]:
import pyrugga as pgr

Pyrugga requires a Super Scout file containing a play by play descriptions of a match. They are stored in an XML format which is not great for statistical analysis, but useful for other things. We need to convert the XML format  into something a little more friendly, a Pandas Dataframes. 

In [52]:
df = pgr.Match('game_1.xml')

A Match object contains a number of functions and Dataframes to help us analyse a match 

**Dataframes**

* Summary -- Summary of the match 
* Events -- Description of each action 
* Timeline -- A timeline of the match with periods ending when either possession of the ball changes or there was a stopage in play 

**Functions**

* getRef -- Will return the name of the referee
* Draw -- Will return if the match was drawn ?
* HomeWin -- Will return if home team won ?

to view the summary  line of a match

In [53]:
df.summary

fixture_code  ref_id ref_name fixture_date  fx_week      awayteam  \
0        719101     204   Peyper   27/10/2018        9  Natal Sharks   

           hometeam  home_score  away_score  
0  Western Province          12          17

to access the first 10 events of a match 

In [54]:
df.events.head(10)

action_id       additional  advantage       description       event  \
0     18665513              NaN          0               NaN      Period   
0     18665514              NaN          0               NaN     Restart   
94    18665515              NaN          0               NaN  Collection   
65    18665516              NaN          0               NaN    Turnover   
98    18665517              NaN          0               NaN  Collection   
218   18665518  Neutral Contact          0           Neutral       Carry   
404   18665519              NaN          0               NaN      Tackle   
137   18665520              NaN          0               NaN        Ruck   
680   18665521              NaN          0               NaN        Pass   
231   18665522  Neutral Contact          0  Crossed Gainline       Carry   

                event_type  fixture_code  home_team_advantage  match_time  \
0             Start Period      719101.0                    0           0   
0         50m Restart Kick      719101.0                    0           0   
94           Restart Catch      719101.0                    0           4   
65   Dropped Ball Unforced      719101.0                    0           4   
98    Defensive Loose Ball      719101.0                    0           5   
218            Other Carry      719101.0                    0           7   
404            Line Tackle      719101.0                    0           7   
137                    NaN      719101.0                    0           7   
680          Complete Pass      719101.0                    0           7   
231          One Out Drive      719101.0                    0          12   

     metres  ... ps_endstamp  ps_timestamp  set_num shirt_no  \
0         0  ...           0             0        0      0.0   
0         0  ...           0             0        0     15.0   
94        0  ...           4             4        1     11.0   
65        0  ...           4             4        1     11.0   
98        0  ...           5             5        1      6.0   
218       0  ...           7             7        1      6.0   
404       0  ...           7             7        1     11.0   
137       0  ...           7             7        1      0.0   
680       0  ...           7             7        1      9.0   
231       2  ...          12            12        1      3.0   

            team_name  x_coord  x_coord_end  y_coord  y_coord_end points  
0    Western Province       50            0       34            0      0  
0        Natal Sharks       50           75       34           57      0  
94   Western Province       26            0       13            0      0  
65   Western Province       26            0       13            0      0  
98       Natal Sharks       76            0       54            0      0  
218      Natal Sharks       76            0       54            0      0  
404  Western Province       24            0       14            0      0  
137      Natal Sharks       75            0       52            0      0  
680      Natal Sharks       75            0       52            0      0  
231      Natal Sharks       75            0       52            0      0  

[10 rows x 25 columns]

to access the timeline of a match type

In [55]:
df.timeline

period  set_num         team_name  points  x_coord  x_coord_end  y_coord  \
0        1        1      Natal Sharks       0       74           75       55   
1        1        2      Natal Sharks       0       24           59        1   
2        1        3  Western Province       0       41           69       14   
3        1        4      Natal Sharks       0       31          102       21   
4        1        5      Natal Sharks       0       40           50       51   
5        1        6  Western Province       0       76           97        3   
6        1        7  Western Province       0       73           92       29   
7        1        8      Natal Sharks       0        4           38       54   
8        1        9      Natal Sharks       0       21           20       64   
9        1       10      Natal Sharks       0       49           89       66   
10       1       11  Western Province       0       11           37       13   
11       1       12  Western Province       0       22           22       15   
12       1       13  Western Province       0       21           62       24   
13       1       14      Natal Sharks       0       38           77       51   
14       1       15  Western Province       0       23           72       21   
15       1       16      Natal Sharks       0       28           49       23   
16       1       17  Western Province       0       85           77       68   
17       1       18      Natal Sharks       0       24           79       22   
18       1       19  Western Province       0       21           85       37   
19       1       20      Natal Sharks      -3       14           33       65   
20       1       21      Natal Sharks       0       65           64       49   
21       1       22  Western Province       0       37           82       46   
22       1       23      Natal Sharks       0       18           61       10   
23       1       24  Western Province       0       39           58       35   
24       1       25      Natal Sharks       0       42           48       35   
25       1       26      Natal Sharks       0       42           91       46   
26       1       27  Western Province       0       18           67       15   
27       1       28      Natal Sharks       0       33           62       65   
28       1       29  Western Province       0       80           77       67   
29       1       30      Natal Sharks       0       21           17       31   
..     ...      ...               ...     ...      ...          ...      ...   
51       2       52      Natal Sharks       0       10           33       67   
52       2       53  Western Province       0       67           58        0   
53       2       54      Natal Sharks       0       42           63       52   
54       2       55  Western Province       0       37           64       21   
55       2       56      Natal Sharks       0       36           34       18   
56       2       57  Western Province       0       66           92       50   
57       2       58      Natal Sharks       0        8           36       14   
58       2       59  Western Province       0       62           59       65   
59       2       60      Natal Sharks       0       41           85       47   
60       2       61  Western Province       0       15           46        3   
61       2       62      Natal Sharks       0       53           54       68   
62       2       63      Natal Sharks       0       42           76       57   
63       2       64  Western Province       0       24           37       30   
64       2       65      Natal Sharks       0       65           64       51   
65       2       66  Western Province       0       36           31       23   
66       2       67      Natal Sharks       0       69           71       56   
67       2       68  Western Province       0       80           78        0   
68       2       69      Natal Sharks       0       21           44       47   
69  

And so on ...

In [56]:
df.Draw()

False

In [57]:
df.HomeWin()

False

In [58]:
df.getRef()

'Peyper'

# Some Analysis

On to something a little more useful such as working out working out which team had the most possession during a game. Our match is still stored in variable **df** and it contains a Dataframe called timeline. To access the timeline 

```python
df.timeline
```

then we will want to group by team_name and sum up the time each team had the ball. Group by is simple enough just .groupby('team_name') 

```python
df.timeline.groupby('team_name') 
```

and then to sum add .sum() and for length add ['length'] leaving us with 

In [59]:
df.timeline.groupby('team_name').sum()['length'] 

team_name
Natal Sharks        1349
Western Province     707
Name: length, dtype: int64

We can see the Sharks has a 1349 seconds with the ball in play compare to Province 707 secounds. What if we wanted to see this in percentages we need to sum the column which requires adding .sum() 


In [60]:
df.timeline.groupby('team_name').sum()['length'] / (df.timeline.groupby('team_name').sum()['length'].sum())

team_name
Natal Sharks        0.656128
Western Province    0.343872
Name: length, dtype: float64

Say you want to see points per time with ball. We can see Province was slightly more effective scoring a point on average every 78 secs with the ball compared to Natal with 96secs

In [61]:
df.timeline.groupby(['team_name']).sum()['length'] / (df.timeline.groupby('team_name')['points'].sum())

team_name
Natal Sharks        96.357143
Western Province    78.555556
dtype: float64

What about that other TV stat, territory I here you say. We need a variable called zone. For that we are better off looking at indiviudal actions in the event Dataframe. 

In [65]:
pos = df.events.groupby(['team_name','x_coord']).count()['action_id']

In [68]:
pos.reset_index()['team_']

team_name  x_coord  action_id
0        Natal Sharks       -5          1
1        Natal Sharks        3          5
2        Natal Sharks        4          2
3        Natal Sharks        6          2
4        Natal Sharks        8          5
5        Natal Sharks        9          6
6        Natal Sharks       10          7
7        Natal Sharks       11          1
8        Natal Sharks       12          7
9        Natal Sharks       13          2
10       Natal Sharks       14          3
11       Natal Sharks       15          1
12       Natal Sharks       18          2
13       Natal Sharks       19         12
14       Natal Sharks       20          7
15       Natal Sharks       21         17
16       Natal Sharks       22          7
17       Natal Sharks       23          2
18       Natal Sharks       24         11
19       Natal Sharks       25         14
20       Natal Sharks       26          6
21       Natal Sharks       27          5
22       Natal Sharks       28         13
23       Natal Sharks       29          7
24       Natal Sharks       30          4
25       Natal Sharks       31          2
26       Natal Sharks       32          8
27       Natal Sharks       33          5
28       Natal Sharks       34          2
29       Natal Sharks       36          3
..                ...      ...        ...
146  Western Province       58         14
147  Western Province       59         10
148  Western Province       60          1
149  Western Province       61          5
150  Western Province       62          5
151  Western Province       64          3
152  Western Province       65          4
153  Western Province       66          2
154  Western Province       67          5
155  Western Province       69          1
156  Western Province       71          2
157  Western Province       72          4
158  Western Province       73         14
159  Western Province       74         11
160  Western Province       75          6
161  Western Province       76         18
162  Western Province       77          3
163  Western Province       78         10
164  Western Province       79          7
165  Western Province       80         15
166  Western Province       81         12
167  Western Province       82          3
168  Western Province       85          5
169  Western Province       86          2
170  Western Province       88          6
171  Western Province       89          2
172  Western Province       91          3
173  Western Province       95          2
174  Western Province       96          2
175  Western Province       97          1

[176 rows x 3 columns]

In [16]:
list(df.timeline.columns)

['period',
 'set_num',
 'team_name',
 'points',
 'x_coord',
 'x_coord_end',
 'y_coord',
 'y_coord_end',
 'meters_gained',
 'dist_traveled',
 'start',
 'length',
 'start_event',
 'end_event',
 'phases',
 'carry',
 'collection',
 'other_carry',
 'kick_return',
 'one_out_drive',
 'defender_beaten',
 'defensive_catch',
 'defensive_loose_ball',
 'attacking_loose_ball',
 'support_carry',
 'pick_and_go',
 'stepped',
 'dropped_ball_unforced',
 'initial_break',
 'restart_catch',
 'restart_return',
 'supported_break',
 'kick',
 'goal_kick',
 'box',
 'territorial',
 'bomb',
 'touch_kick',
 'lineout',
 'lineout_take',
 'throw_middle',
 'lineout_win_middle',
 'lineout_win_front',
 'maul',
 'ruck',
 'pass',
 'complete_pass',
 'offload',
 'scrum_half_pass',
 'break_pass',
 'incomplete_pass',
 'penalty_conceded',
 'penalty_won',
 'scrum',
 'defensive_scrum',
 'offensive_scrum',
 'no_8_pick_up',
 'no_8_pass',
 'tackle',
 'missed_tackle',
 'line_tackle',
 'chase_tackle',
 'other_tackle',
 'cover_tackle'